### ISSUE TITLE
unexpected include_task behaviour

### ISSUE THREAD

**Issue by philfry** [ansible/ansible#31409]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME

core

##### ANSIBLE VERSION
```
ansible 2.5.0
  config file = /home/x/projects/tmp/ansiblebug/ansible.cfg
  configured module search path = [u'/home/x/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/x/projects/tmp/ansible_master/lib/python2.7/site-packages/ansible
  executable location = /home/x/projects/tmp/ansible_master/bin/ansible
  python version = 2.7.13 (default, Sep  5 2017, 08:53:59) [GCC 7.1.1 20170622 (Red Hat 7.1.1-3)]
```

##### CONFIGURATION
```
ANSIBLE_NOCOWS(/home/x/projects/tmp/ansiblebug/ansible.cfg) = True
DEFAULT_LOG_PATH(env: ANSIBLE_LOG_PATH) = /home/x/.ansible/log
DISPLAY_SKIPPED_HOSTS(/home/x/projects/tmp/ansiblebug/ansible.cfg) = False
RETRY_FILES_ENABLED(/home/x/projects/tmp/ansiblebug/ansible.cfg) = False
```

##### OS / ENVIRONMENT
n/a

##### SUMMARY
A conditional `include_tasks` is not running tasks, even though I think it should.

Back story: I have several roles that can be included either a standalone or a dependent role (like a webserver, a database role, ...) and I don't want ansible to run such a role or parts of such a role multiple times. That's why I set a variable at the end of such a role to mark it as completed, and the variable is checked at the beginning of the role.

When using `import_tasks` instead of `include_tasks` in `dependency-or-standalone/tasks/main.yml` I get the "expected" results, but a) in my understanding -- please correct me, if I'm wrong -- of the difference between statically and dynamically included tasks, this should be a dynamic include and b) using `import_tasks` would result in every task of non-relevant roles being evaluated at playbook-compile-time. This increases runtime dramatically on playbooks with many roles/tasks and might lead to errors when referring to facts that do not exist at compile-time yet.

##### STEPS TO REPRODUCE
```yaml
git clone https://github.com/philfry/myansibleissue01
cd myansibleissue01
ansible-play site.yml
```

##### EXPECTED RESULTS

```
[...]
TASK [dependency-or-standalone : completed roles before run] *************************************************************************************************************************
ok: [localhost] => {
    "dep_completed": "VARIABLE IS NOT DEFINED!"
}

TASK [dependency-or-standalone : debug] **********************************************************************************************************************************************
ok: [localhost] => {
    "msg": "in dependency-or-standalone"
}

TASK [dependency-or-standalone : set_fact] *******************************************************************************************************************************************
ok: [localhost]

TASK [dependency-or-standalone : completed roles after run] **************************************************************************************************************************
ok: [localhost] => {
    "dep_completed": true
}

TASK [role2 : debug] *****************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": "in master"
}
[...]
```
##### ACTUAL RESULTS
missing:
```
ok: [localhost] => {
    "msg": "in dependency-or-standalone"
}
```
full output:
```
ansible-playbook 2.5.0
  config file = /home/x/projects/tmp/ansiblebug/ansible.cfg
  configured module search path = [u'/home/x/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/x/projects/tmp/ansible_master/lib/python2.7/site-packages/ansible
  executable location = /home/x/projects/tmp/ansible_master/bin/ansible-playbook
  python version = 2.7.13 (default, Sep  5 2017, 08:53:59) [GCC 7.1.1 20170622 (Red Hat 7.1.1-3)]
Using /home/x/projects/tmp/ansiblebug/ansible.cfg as config file
setting up inventory plugins
Parsed /etc/ansible/hosts inventory source with yaml plugin
Loading callback plugin default of type stdout, v2.0 from /home/x/projects/tmp/ansible_master/lib/python2.7/site-packages/ansible/plugins/callback/__init__.pyc

PLAYBOOK: site.yml *******************************************************************************************************************************************************************
1 plays in site.yml

PLAY [localhost] *********************************************************************************************************************************************************************
META: ran handlers

TASK [dependency-or-standalone : completed roles before run] *************************************************************************************************************************
task path: /home/x/projects/tmp/ansiblebug/roles/dependency-or-standalone/tasks/main.yml:2

TASK [dependency-or-standalone : include_tasks] **************************************************************************************************************************************
task path: /home/x/projects/tmp/ansiblebug/roles/dependency-or-standalone/tasks/main.yml:6

TASK [dependency-or-standalone : set_fact] *******************************************************************************************************************************************
task path: /home/x/projects/tmp/ansiblebug/roles/dependency-or-standalone/tasks/main.yml:7

TASK [dependency-or-standalone : completed roles after run] **************************************************************************************************************************
task path: /home/x/projects/tmp/ansiblebug/roles/dependency-or-standalone/tasks/main.yml:10

TASK [dependency-or-standalone : completed roles before run] *************************************************************************************************************************
task path: /home/x/projects/tmp/ansiblebug/roles/dependency-or-standalone/tasks/main.yml:2

TASK [dependency-or-standalone : include_tasks] **************************************************************************************************************************************
task path: /home/x/projects/tmp/ansiblebug/roles/dependency-or-standalone/tasks/main.yml:6

TASK [dependency-or-standalone : set_fact] *******************************************************************************************************************************************
task path: /home/x/projects/tmp/ansiblebug/roles/dependency-or-standalone/tasks/main.yml:7

TASK [dependency-or-standalone : completed roles after run] **************************************************************************************************************************
task path: /home/x/projects/tmp/ansiblebug/roles/dependency-or-standalone/tasks/main.yml:10

TASK [role1 : debug] *****************************************************************************************************************************************************************
task path: /home/x/projects/tmp/ansiblebug/roles/role1/tasks/main.yml:2

TASK [dependency-or-standalone : completed roles before run] *************************************************************************************************************************
task path: /home/x/projects/tmp/ansiblebug/roles/dependency-or-standalone/tasks/main.yml:2
ok: [localhost] => {
    "dep_completed": "VARIABLE IS NOT DEFINED!"
}

TASK [dependency-or-standalone : include_tasks] **************************************************************************************************************************************
task path: /home/x/projects/tmp/ansiblebug/roles/dependency-or-standalone/tasks/main.yml:6
included: /home/x/projects/tmp/ansiblebug/roles/dependency-or-standalone/tasks/main_run.yml for localhost

TASK [dependency-or-standalone : debug] **********************************************************************************************************************************************
task path: /home/x/projects/tmp/ansiblebug/roles/dependency-or-standalone/tasks/main_run.yml:2

TASK [dependency-or-standalone : set_fact] *******************************************************************************************************************************************
task path: /home/x/projects/tmp/ansiblebug/roles/dependency-or-standalone/tasks/main.yml:7
ok: [localhost] => {
    "ansible_facts": {
        "dep_completed": true
    }, 
    "changed": false, 
    "failed": false
}

TASK [dependency-or-standalone : completed roles after run] **************************************************************************************************************************
task path: /home/x/projects/tmp/ansiblebug/roles/dependency-or-standalone/tasks/main.yml:10
ok: [localhost] => {
    "dep_completed": true
}

TASK [role2 : debug] *****************************************************************************************************************************************************************
task path: /home/x/projects/tmp/ansiblebug/roles/role2/tasks/main.yml:2
ok: [localhost] => {
    "msg": "in master"
}
META: ran handlers
META: ran handlers

PLAY RECAP ***************************************************************************************************************************************************************************
localhost                  : ok=5    changed=0    unreachable=0    failed=0   
```

**Comment by philfry** [comment]
fwiw: by replacing
```yaml
roles:
  - role: whatever
```
with
```yaml
tasks:
  - include_role: name=whatever
```
I get the expected results.

**Comment by sivel** [comment]
To be honest here, I'm having a hard time determining what your concern/question/problem really is.

You've effectively not provided a playbook, and just listed a big chunk of outputs expected and actual, and vaguely indicated there is a difference between `import_tasks` and `include_tasks`

**Comment by philfry** [comment]
@sivel please see "steps to reproduce". The playbook is in the stated git repository.

**Comment by sivel** [comment]
Ok, that answers the no playbook question, I still can't determine what your concern/question/problem really is.

**Comment by philfry** [comment]
Thanks for your reply.
My concern/question/problem is that ansible acts weird.

The playbook contains a couple of roles that get included on a conditional state. In this case, only `role2` should run, because the array `server_roles` only contains `role2`.

The roles `role1` and `role2` have the `dependency-or-standalone` role as dependency. So running `role2` should run `dependency-or-standalone`'s tasks, right? But it doesn't.

**To make it even creepier**, remove `role1` from the playbook (in the git repo, I created a second playbook named `site-working.yml`). We should expect the same behaviour as in `site.yml`, right? No: the role `dependency-or-standalone`'s tasks get executed this time. Can anyone explain that?

I found two workarounds:
- `site-workaround1.yml` uses `tasks` and `import_role` instead of `roles` and `role`
- `site-workaround2.yml` uses `import_tasks` instead of `include_tasks` but leads to compile-time evaluation of roles that should never run.

Workaround -- fine. But in my opinion ansible is broken here and should behave differently.

The problem reminds me somehow of #23733, where conditionals broke role imports.

**Comment by philfry** [comment]
To (hopefully) make it a little clearer, I rewrote my git repo, added a little explanation of what my intention is, and moved the playbook variants to different branches: https://github.com/philfry/myansibleissue01

The branch `master` contains the playbook as it should be correct for ansible 2.4, the branch `ansible-2.3` contains the old-style playbook with `include:` instead of `include_tasks:`. The playbook run output is included.
Please check out my repo and take a look at
```
git diff ansible-2.3 master
git diff ansible-2.3 ansible-2.4-import_tasks
git diff ansible-2.3 ansible-2.4-include_role
git diff ansible-2.3 ansible-2.4-removed-role
```
Thanks.

**Comment by sivel** [comment]
This issue was resolved in f1c458303a1ad00675497597e1f7fbc0539e314e and included in the v2.4.3 release.

If you have further questions please stop by IRC or the mailing list:

   * IRC: #ansible on irc.freenode.net
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project
